### ISSUE TITLE
Allow user_agent to take *args which are tuples

### ISSUE THREAD

**Issue by sigmavirus24** [sigmavirus24/requests-toolbelt#102]
twine is going to start [sending its own user-agent string](https://github.com/pypa/twine/issues/139) and I'd like to not only report the version of twine but also the versions of its dependencies.

Usage would look like:

```py
user_agent('twine', '1.7.0', ('pkginfo', pi_version), ...)
```

And it should assert that the tuples provided are of length 2.

**Comment by sigmavirus24** [comment]
Actually, I'm thinking this is better without using `*args`, e.g.,

```py
user_agent('twine', '1.7.0', extras=[('pkginfo', pi_version), ('requests', '2.8.0'), ...])
```

This also allows for a flag to enable/disable sending platform information.

**Comment by sigmavirus24** [comment]
So actually, in refreshing my memory of the spec, each of these can have a comment, e.g.,

```
product/product-version (comment) product/product-version (comment)
```

Etc.

I'm thinking that we should re-design this to at least use `namedtuples` in the implementation but to also accept them. That said, that breaks the signature of the existing `user_agent` function. I think I'll have to add a new function that takes the `namedtuple`\ s and turns them into a user agent string.

**Comment by Lukasa** [comment]
Eh, they *can* have comments, but what is the value in supporting them?

**Comment by sigmavirus24** [comment]
*shrug* It probably wouldn't hurt to allow for the slight extra bits that the spec pieces allow for. I don't know if twine will ever use it, but it could eventually become worthwhile

**Comment by sigmavirus24** [comment]
And in hindsight if it ever does become worthwhile I can rewrite this stuff.

**Comment by sigmavirus24** [comment]
Merged with #104
